### PR TITLE
Change the behavior of formatter %M in function formatDateTime() from minutes to month name

### PR DESCRIPTION
**Pull request by rschu1ze**
With this PR, MySQL format string `%M` in SQL functions `parseDateTime()` and `formatDateTime()` now matches [mySQL's behavior](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_date-format) which returns the full month name (e.g. February) instead of minutes (e.g. 00-59).

This finishes #46536. Courtesy to @JakeBamrah 

As a side effect, this also fixes asan failure #47291 and it's duplicates. Added label 'pr-must-backport' for that reason.

Review note:
- The PR Includes two cleanup commit which should be reviewed separately
- The actual changes are much smaller. Changes in `formatDateTime()` are a bit more complicated as the infrastructure to generate variable-length output was lacking

The fix is technically backward-incompatible (--> setting label "pr-backward-incompatible"), I added a setting to restore the old behavior if wanted. The backports will have the settings configured to the old behavior by default.

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md)
Formatter '%M' in function formatDateTime() now prints the month name instead of the minutes. This makes the behavior consistent with MySQL. The previous behavior can be restored using setting "formatdatetime_parsedatetime_m_is_month_name = 0".

### CODE REVIEW

**Comment by rschu1ze** [comment]
ClickHouse Stateless Tests (debug, S3 Storage) [3/6]:
- 02232_dist_insert_send_logs_level_hung: was flaky in the last three days

ClickHouse Upgrade Check (debug):
- errors caused by OOM killer

**Comment by JakeBamrah** [comment]
The work looks great! Well done 😊

**Comment by rschu1ze** [comment]
Thanks 😄 

Let's put this PR on hold until #46815 is in (there are logical + physical merge conflicts). The same change in behavior of `%M` has to be applied to `parseDateTime()`.

**Comment by rschu1ze** [comment]
> Let's put this PR on hold until #46815 is in (there are logical + physical merge conflicts). The same change in behavior of `%M` will have to be applied to `parseDateTime()`.

Nope, let's try to merge this PR asap as it fixes a memory corruption. I added a commit "Cleanup, pt. II" on top which should keep merge conflicts with @taiyang-li's  #46815 as minimal as possible.

**Comment by rschu1ze** [comment]
Over 70 merge conflicts after #46815 went in 😢

File: src/Functions/formatDateTime.cpp [hunk 1]
```diff
@@ -38,22 +38,19 @@ namespace ErrorCodes
 
 namespace
 {
+using Pos = const char *;
 
-struct FormatDateTimeTraits
+enum class SupportInteger
```
Comment (antonio2368) [inline src/Functions/formatDateTime.cpp:43]:
```suggestion
enum class SupportInteger : uint8_t
```

File: src/Functions/formatDateTime.cpp [hunk 2]
```diff
@@ -38,22 +38,19 @@ namespace ErrorCodes
...
-    enum class FormatSyntax
-    {
-        MySQL,
-        Joda
-    };
+    Yes,
+    No
 };
 
+enum class FormatSyntax
```
Comment (antonio2368) [inline src/Functions/formatDateTime.cpp:49]:
```suggestion
enum class FormatSyntax : uint8_t
```

File: src/Functions/formatDateTime.cpp [hunk 3]
```diff
@@ -152,26 +147,34 @@ class FunctionFormatDateTimeImpl : public IFunction
...
+
+        void setMysqlFunc(FuncMysql && func) { func_mysql = std::move(func); }
+        void setJodaFunc(FuncJoda && func) { func_joda = std::move(func); }
+        void setLiteral(std::string_view literal_) { literal = literal_; }
 
         void perform(char *& dest, Time source, UInt64 fractional_second, UInt32 scale, const DateLUTImpl & timezone)
         {
-            auto shift = func(dest, source, fractional_second, scale, timezone);
+            size_t shift = func_mysql
+                           ? ((static_cast<Instruction<Time>*>(this))->*func_mysql)(dest, source, fractional_second, scale, timezone)
```
Comment (antonio2368) [inline src/Functions/formatDateTime.cpp:176]: Wow...

Why do we have to cast this?

File: src/Functions/formatDateTime.cpp [hunk 4]
```diff
@@ -651,6 +667,34 @@ class FunctionFormatDateTimeImpl : public IFunction
...
     };
 
+    [[noreturn]] static void throwLastCharacterIsPercentException()
+    {
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "'%' must not be the last character in the format string, use '%%' instead");
+    }
+
+    static bool containsOnlyFixedWidthMySQLFormatters(std::string_view format)
+    {
+        constexpr std::array variable_width_formatter = {'M', 'W'};
```
Comment (antonio2368) [inline src/Functions/formatDateTime.cpp:677]:
```suggestion
        static constexpr std::array variable_width_formatter = {'M', 'W'};
```

File: src/Functions/formatDateTime.cpp [hunk 5]
```diff
@@ -790,10 +834,23 @@ class FunctionFormatDateTimeImpl : public IFunction
         if constexpr (std::is_same_v<DataType, DataTypeDateTime64>)
             scale = times->getScale();
 
+        /// For MySQL, we support two modes of execution:
+        /// - All formatters in the format string are fixed-width. As a result, all output rows will have the same with and structure. We
```
Comment (antonio2368) [inline src/Functions/formatDateTime.cpp:838]:
```suggestion
        /// - All formatters in the format string are fixed-width. As a result, all output rows will have the same width and structure. We
```

File: src/Functions/formatDateTime.cpp [hunk 6]
```diff
@@ -790,10 +834,23 @@ class FunctionFormatDateTimeImpl : public IFunction
...
+        ///   take advantage of this and 1. create a "template" with placeholders from the format string, 2. allocate a result column large
+        ///   enough to store the template on each row, 3. copy the template into each result row 4. run instructions which replace the
+        ///   formatter placeholders. All other parts of the template (e.g. whitespaces) are already as desired and instructions skip over
+        ///   them (see 'extra_shift' in the formatters).
```
Comment (antonio2368) [inline src/Functions/formatDateTime.cpp:842]: split steps into rows, hard to read

File: src/Functions/formatDateTime.cpp [hunk 7]
```diff
@@ -855,289 +910,445 @@ class FunctionFormatDateTimeImpl : public IFunction
...
-                        add_instruction_or_extra_shift(&Instruction<T>::mysqlMinute, 2);
-                        out_template += "00";
-                        break;
-
                     // AM or PM
                     case 'p':
-                        add_instruction_or_extra_shift(&Instruction<T>::mysqlAMPM, 2);
-                        out_template += "AM";
+                    {
+                        constexpr std::string_view val = "AM";
```
Comment (antonio2368) [inline src/Functions/formatDateTime.cpp:1223]:
```suggestion
                        static constexpr std::string_view val = "AM";
```

Comment (rschu1ze) [reply src/Functions/formatDateTime.cpp:1223]: done, tnx

File: src/Functions/formatDateTime.cpp (hunk 6 above)
Comment (rschu1ze) [reply src/Functions/formatDateTime.cpp:842]: yepp, done

File: src/Functions/formatDateTime.cpp (hunk 5 above)
Comment (rschu1ze) [reply src/Functions/formatDateTime.cpp:838]: done

File: src/Functions/formatDateTime.cpp (hunk 4 above)
Comment (rschu1ze) [reply src/Functions/formatDateTime.cpp:677]: sure, fixed

File: src/Functions/formatDateTime.cpp (hunk 1 above)
Comment (rschu1ze) [reply src/Functions/formatDateTime.cpp:43]: Not really sure about this. There are cases where the specification of the underlying type is crucial, e.g. if the enum values are persisted. Here, enum values merely exist at compile-time, and an explicit type is visual clutter, imho.

File: src/Functions/formatDateTime.cpp (hunk 2 above)
Comment (rschu1ze) [reply src/Functions/formatDateTime.cpp:49]: (see above)

File: docs/en/sql-reference/functions/date-time-functions.md [hunk 8]
```diff
@@ -1276,16 +1276,16 @@ Using replacement fields, you can define a pattern for the resulting string. “
 | %k       | hour in 24h format (00-23)                              | 22         |
 | %l       | hour in 12h format (01-12)                              | 09         |
 | %m       | month as an integer number (01-12)                      | 01         |
-| %M       | minute (00-59)                                          | 33         |
+| %M       | full month name (January-December)                      | January    |
```
[the remark this answers is not in the document]
Comment (rschu1ze) [reply docs/en/sql-reference/functions/date-time-functions.md:1279]: The PR message says "The PR can be viewed either as bugfix or a backwards incompatible change. I'll set the "pr-backward-incompatible" tag for now." 😄 I am not really sure, I'll ask Nikolai later.

[the remark this answers is not in the document]
Comment (antonio2368) [reply docs/en/sql-reference/functions/date-time-functions.md:1279]: I see you're having a great battle with the bots when trying to add the tag 😄

File: src/Functions/formatDateTime.cpp [hunk 9]
```diff
@@ -807,26 +868,26 @@ class FunctionFormatDateTimeImpl : public IFunction
...
-        auto & dst_offsets = col_res->getOffsets();
-        dst_data.resize(vec.size() * (result_size + 1));
-        dst_offsets.resize(vec.size());
+        auto & res_data = col_res->getChars();
+        auto & res_offsets = col_res->getOffsets();
+        res_data.resize(vec.size() * (out_template_size + 1));
+        res_offsets.resize(vec.size());
 
-        if constexpr (format_syntax == FormatDateTimeTraits::FormatSyntax::MySQL)
+        if (format_syntax == FormatSyntax::MySQL && only_fixed_width_formatters)
```
Comment (antonio2368) [inline src/Functions/formatDateTime.cpp:876]: I'd rather extra indentation to keep compile time `if`

File: src/Functions/formatDateTime.cpp [hunk 10]
```diff
@@ -855,289 +914,445 @@ class FunctionFormatDateTimeImpl : public IFunction
...
             instructions.back().extra_shift += amount;
         };
 
-        auto add_instruction_or_extra_shift = [&](auto * func [[maybe_unused]], size_t amount [[maybe_unused]])
+        auto add_literal_instruction = [&](std::string_view literal)
+        {
+            Instruction<T> instruction;
+            instruction.setMysqlFunc(&Instruction<T>::mysqlLiteral);
+            instruction.setLiteral(literal);
+            instructions.push_back(instruction);
```
Comment (antonio2368) [inline src/Functions/formatDateTime.cpp:954]:
```suggestion
            instructions.push_back(std::move(instruction));
```

same for all `push_back(instruction)`

File: src/Functions/formatDateTime.cpp [hunk 11]
```diff
@@ -1172,32 +1386,41 @@ class FunctionFormatDateTimeImpl : public IFunction
...
 
     template <typename T>
-    size_t parseJodaFormat(const String & format, std::vector<Instruction<T>> & instructions, UInt32, String &) const
+    size_t parseJodaFormat(const String & format, std::vector<Instruction<T>> & instructions, UInt32, bool, String &) const
     {
         /// If the argument was DateTime, add instruction for printing. If it was date, just append default literal
-        auto add_instruction = [&](auto && func [[maybe_unused]], const String & default_literal [[maybe_unused]])
+        auto add_instruction = [&]([[maybe_unused]] typename Instruction<T>::FuncJoda && func, [[maybe_unused]] const String & default_literal)
         {
             if constexpr (std::is_same_v<T, UInt32> || std::is_same_v<T, Int64>)
```
Comment (antonio2368) [inline src/Functions/formatDateTime.cpp:1394]: maybe move `concept is_any_of` from `typeid_cast.h` to `Common/Concepts.h` and use it here?

**Review by antonio2368 — APPROVED** [review]
LGTM, the only question being backwards compatiblity

File: src/Functions/formatDateTime.cpp (hunk 9 above)
Comment (rschu1ze) [reply src/Functions/formatDateTime.cpp:876]: Makes sense, fixed.

File: src/Functions/formatDateTime.cpp (hunk 10 above)
Comment (rschu1ze) [reply src/Functions/formatDateTime.cpp:954]: Yepp, done.

File: src/Functions/formatDateTime.cpp (hunk 11 above)
Comment (rschu1ze) [reply src/Functions/formatDateTime.cpp:1394]: Sounds like a good idea (didn't know about `is_any_of`), done.

**Comment by rschu1ze** [comment]
> LGTM, the only question being backwards compatiblity

Will add a backward compat setting ...

**Comment by rschu1ze** [comment]
The PR should be in a mergeable state now (fingers crossed ...)

**Comment by den-crane** [comment]
I would do 

```
formatdatetime_parsedatetime_m_is_month_name=throw (default)
formatdatetime_parsedatetime_m_is_month_name=mysql_compat
formatdatetime_parsedatetime_m_is_month_name=legacy_compat
```

**Comment by rschu1ze** [comment]
@den-crane With the benefit of hindsight, I agree. On the other hand, this PR appeared as a backward-incompatible change in the changelog (v23.4, [here](https://clickhouse.com/docs/en/whats-new/changelog)) and the new behavior is well documented ("Note 1", [here](https://clickhouse.com/docs/en/sql-reference/functions/date-time-functions#date_time_functions-formatDateTime)). Changing the setting once more would create unnecessary confusion, IMHO. And at some point (1-2 years from now?), we probably might want to get rid of `formatdatetime_parsedatetime_m_is_month_name` anyways.

**Pull request merged**